relay-pattern/src/wildmatch.rs

```diff
@@ -0,0 +1,605 @@
+use std::num::NonZeroUsize;
+
+use crate::{Literal, Options, Ranges, Token, Tokens};
+
+/// Matches [`Tokens`] against a `haystack` with the provided [`Options`].
+///
+/// This implementation is largely based on the algorithm described by [Kirk J Krauss]
+/// and combining the two loops into a single one and other small modifications to take advantage
+/// of the already pre-processed [`Tokens`] structure and its invariants.
+///
+/// [Kirk J Krauss]: http://developforperformance.com/MatchingWildcards_AnImprovedAlgorithmForBigData.html
+pub fn is_match(tokens: &Tokens, haystack: &str, options: Options) -> bool {
```

Comment:
nit: Rename `tokens` to `needle` for consistency with `haystack`.

Reply:
I would like to keep `tokens` here, this is also used in the `is_match_impl` in various places (e.g. the `t_*` variables) and keep the `needle` and `haystack` terminology to strings. This also separates logic when you're matching a token and searching for a string/character.